### DISCUSSION TITLE
`enable_input_require_grads` is called twice with `use_gradient_checkpointing=True` and bnb quantisation; is there any effect?

### DISCUSSION THREAD

**Discussion by nzw0301** [huggingface/peft#1623] [Q&A] [answered]
As in [the quantization tutorial](https://github.com/huggingface/peft/blob/8452d71e14f1d53a1ad580a3c57566dd4f59c6ea/docs/source/developer_guides/quantization.md?plain=1#L59), we need to call the following lines to use peft with quantization.

```python
from peft import prepare_model_for_kbit_training
model = prepare_model_for_kbit_training(model)
```

In the method with `use_gradient_checkpointing=True`, which is a default value,
`model.enable_input_require_grads` is called in https://github.com/huggingface/peft/blob/8452d71e14f1d53a1ad580a3c57566dd4f59c6ea/src/peft/utils/other.py#L111-L122.

Then almost same logic is executed in [`_prepare_model_for_gradient_checkpointing`](https://github.com/huggingface/peft/blob/8452d71e14f1d53a1ad580a3c57566dd4f59c6ea/src/peft/peft_model.py#L443-L460) via `get_peft_model`. So in my understanding, `model.enable_input_require_grads` can be called twice.

**Comment by BenjaminBossan** [comment, marked as answer]
Yes, calling this twice should not change the result.

**Reply by nzw0301** [reply]
Thank you for your response!

**Reply by divisionblur** [reply]
What is the function of the line of code `model.enable_input_require_grads` ?

**Reply by divisionblur** [reply]
Is it also necessary to set this during the pre-training phase?

**Reply by BenjaminBossan** [reply]
The explanation can be found here:

https://github.com/huggingface/transformers/blob/0f67ba1d741d65b07d549daf4ee157609ce4f9c1/src/transformers/modeling_utils.py#L2273-L2278

For pre-training, you would typically not use PEFT (instead updating all parameters) so it should not be required.
